### DISCUSSION TITLE
What does the 'distance' argument do in the mesh generation?

### DISCUSSION THREAD

**Discussion by duarte-jfs** [HelgeGehring/femwell#149] [Q&A] [answered]
I am following the tutorials to recreate some results from an RF coplanar waveguide and I am stuck on what the argument "distance" does, as I can never see any changes to the mesh. Can anyone clarify? 

Maybe we can use the code snippet for the rectangular waveguide as an example:

```python
from collections import OrderedDict

import matplotlib.pyplot as plt
import numpy as np
import shapely
import shapely.affinity
from scipy.constants import epsilon_0, speed_of_light
from shapely.ops import clip_by_rect
from skfem import Basis, ElementTriP0
from skfem.io.meshio import from_meshio

from femwell.maxwell.waveguide import compute_modes
from femwell.mesh import mesh_from_OrderedDict
from femwell.visualization import plot_domains

wg_width = 2.5
wg_thickness = 0.3
core = shapely.geometry.box(-wg_width / 2, 0, +wg_width / 2, wg_thickness)
env = shapely.affinity.scale(core.buffer(5, resolution=8), xfact=0.5)

polygons = OrderedDict(
    core=core,
    box=clip_by_rect(env, -np.inf, -np.inf, np.inf, 0),
    clad=clip_by_rect(env, -np.inf, 0, np.inf, np.inf),
)

resolutions = dict(core={"resolution": 0.03, "distance": 0.5})

mesh = from_meshio(mesh_from_OrderedDict(polygons, resolutions, default_resolution_max=10))
mesh.draw().show()

print(mesh.nelements)
```

if I change the distance from 0.5 to 10 it changes from ~4800 to ~5200 mesh elements, but I don't see any significant changes. When is this parameter useful?

**Comment by HelgeGehring** [comment, marked as answer]
Distance defines over which distance the mesh get's back to the `default_resolution_max`. Therefore, for this example it doesn't make too much difference.

We recently added, that you can also define SizeMax to define the max size after this distance.
Maybe try something like `core={"resolution": 0.03, "distance": 0.5, "SizeMax": 0.04}` and you should see a significant difference which often helps to get what you want :)

We should probably q this further in the docs :thinking:

**Reply by duarte-jfs** [reply]
Indeed! This is a great feature and what I wanted to be able to do. However, I don't fully understand the interplay between those arguments. If we look at the example where I put:

```python
resolutions = dict(core={"resolution": 0.1, "distance": 1, "SizeMax": 0.01})

mesh = from_meshio(mesh_from_OrderedDict(polygons, resolutions, default_resolution_max=100))
```

I get the output below where I have made some comments on what I interpret:

![image](https://github.com/HelgeGehring/femwell/assets/60704674/4491122e-688f-4095-b3b1-70fd402d5e4c)

Now the argument 'distance' is well visible and defined, and I see the three different max_resolution values that we define. However, you mention that 

> We recently added, that you can also define SizeMax to define the max size after this distance.

But it seems that SizeMax defines the resolution exactly at the boundary at a 'distance' of the polygon (outermost rectangle), and not up until that boundary or beyond it. That would explain some 'grading' of the size of the elements up to the boundary. Is this correct?

**Reply by HelgeGehring** [reply]
SizeMax should set the resolution at the boundary. There should be a linear grading from surface of the physical (core) you define over that distance. Beyond that distance, `default_resolution_max` defines the resolution again.

So it makes sense to me what we see here, the core has a resolution of 0.1; over a distance of 1 the resolution increases to 0.01 until after distance the mesh elements get larger as fast as possible.

Usually, I use SizeMax>resolution, as often the physical itself needs to be quite fine and the neighborhood should be rather fine, too, but the further we go away the more coarse it can be.

**Reply by duarte-jfs** [reply]
Now it makes much more sense! Thank you so much, this will be quite useful :)
